File: abc407/d/main.py
from collections import defaultdict, deque
from itertools import combinations, permutations
from bisect import bisect, bisect_left, bisect_right
from sortedcontainers import SortedSet, SortedDict, SortedList
import math
import sys
sys.setrecursionlimit(4100000)
import heapq
try:
    from icecream import ic
except ImportError:  # Graceful fallback if IceCream isn't installed.
    ic = lambda *a: None if not a else (a[0] if len(a) == 1 else a)
# def error(*args, end="\n"): print("[stderr]", *args, end=end, file=sys.stderr)
MOD = 998244353
INF = float("inf")
MINF = -float("inf")

H, W = map(int, input().split())
# A = list(map(int, input().split()))
F = []
for _ in range(H):
    F.append(list(map(int, input().split())))

seen = [[0]*W for _ in range(H)]
d = [(0, 1), (1, 0)]

def calc_xor(seen):
    ret = 0
    for h in range(H):
        for w in range(W):
            if seen[h][w] < 2:
                ret ^= F[h][w]
    return ret

ans = calc_xor(seen)

def dfs(seen):
    global ans
    ans = max(ans, calc_xor(seen))
    for h in range(H):
        for w in range(W):
            if seen[h][w] > 0:
                continue
            seen[h][w] = 1
            for k in range(2):
                nh = h + d[k][0]
                nw = w + d[k][1]
                if 0 <= nh < H and 0 <= nw < W and seen[nh][nw] == 0:
                    seen[nh][nw] = seen[h][w] = 2
                    dfs(seen)
                    seen[nh][nw] = 0
                    seen[h][w] = 1
            dfs(seen)
            seen[h][w] = 0
            return
    
dfs(seen)
print(ans)

File: abc407/d/main.cpp
#include <bits/stdc++.h>
using namespace std;

const int MOD = 998244353;
const int INF = INT_MAX;
const int MINF = INT_MIN;

int H, W;
vector<vector<int>> F;
vector<vector<int>> seen;
int ans = 0;
int dh[2] = {0, 1}; // 右・下
int dw[2] = {1, 0};

// XORを計算する関数
int calc_xor(const vector<vector<int>>& seen) {
    int ret = 0;
    for (int h = 0; h < H; ++h) {
        for (int w = 0; w < W; ++w) {
            if (seen[h][w] == 0) {
                ret ^= F[h][w];
            }
        }
    }
    return ret;
}

void dfs(vector<vector<int>>& seen) {
    ans = max(ans, calc_xor(seen));

    for (int h = 0; h < H; ++h) {
        for (int w = 0; w < W; ++w) {
            if (seen[h][w] > 0) continue;

            seen[h][w] = 1;

            for (int k = 0; k < 2; ++k) {
                int nh = h + dh[k];
                int nw = w + dw[k];
                if (0 <= nh && nh < H && 0 <= nw && nw < W && seen[nh][nw] == 0) {
                    seen[nh][nw] = 1;
                    dfs(seen);
                    seen[nh][nw] = 0;
                }
            }

            seen[h][w] = 0;
        }
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> H >> W;
    F.assign(H, vector<int>(W));
    seen.assign(H, vector<int>(W, 0));

    for (int h = 0; h < H; ++h) {
        for (int w = 0; w < W; ++w) {
            cin >> F[h][w];
        }
    }

    ans = calc_xor(seen);
    dfs(seen);

    cout << ans << '\n';
    return 0;
}
